internal/statemachine/classic_states.go

```diff
@@ -504,84 +513,68 @@ func (stateMachine *StateMachine) manualCustomization() error {
-	/*
-		var classicStateMachine *ClassicStateMachine
-		classicStateMachine = stateMachine.parent.(*ClassicStateMachine)
+// preseedClassicImage calls image.Prepare to seed snaps in classic images
+func (stateMachine *StateMachine) preseedClassicImage() error {
+	var classicStateMachine *ClassicStateMachine
+	classicStateMachine = stateMachine.parent.(*ClassicStateMachine)
 
-		// TODO: Move preseeding logic from livecd-rootfs to ubuntu-image
-		// for all builds
-		if classicStateMachine.Opts.Filesystem != "" &&
-			len(classicStateMachine.commonFlags.Snaps) > 0 {
-			var imageOpts image.Options
+	var imageOpts image.Options
 
-			var err error
-			imageOpts.Snaps, imageOpts.SnapChannels, err = parseSnapsAndChannels(
-				classicStateMachine.commonFlags.Snaps)
-			if err != nil {
-				return err
-			}
-
-			// If the rootfs has already been pre-seeded, we need to delete the
-			// pre-seeded snaps and redo the preseed with all of the snaps
-			stateFile := filepath.Join(classicStateMachine.tempDirs.rootfs,
-				"var", "lib", "snapd", "seed", "seed.yaml")
-			if _, err := os.Stat(stateFile); err == nil {
-				// check for an existing model assertion file, otherwise snapd will use
-				// a generic model assertion
-				modelFile := filepath.Join(classicStateMachine.tempDirs.rootfs,
-					"var", "lib", "snapd", "seed", "assertions", "model")
-				if _, err := os.Stat(modelFile); err == nil {
-					// create a copy of the model file because it will be deleted soon
-					newModelFile := filepath.Join(classicStateMachine.stateMachineFlags.WorkDir,
-						"model")
-					if err := osutilCopyFile(modelFile, newModelFile, 0); err != nil {
-						return fmt.Errorf("Error copying modelFile from preseeded filesystem: %s",
-							err.Error())
-					}
-					imageOpts.ModelFile = newModelFile
-				}
+	var err error
+	imageOpts.Snaps, imageOpts.SnapChannels, err = parseSnapsAndChannels(classicStateMachine.Snaps)
+	if err != nil {
+		return err
+	}
 
-				// Now remove all of the seeded snaps
-				preseededSnaps, err := removePreseeding(
-					classicStateMachine.tempDirs.rootfs)
-				if err != nil {
-					return fmt.Errorf("Error removing preseeded snaps from existing rootfs: %s",
-						err.Error())
-				}
-				for snap, channel := range preseededSnaps {
-					// if a channel is specified on the command line for a snap that was already
-					// preseeded, use the channel from the command line instead of the channel
-					// that was originally used for the preseeding
-					if _, found := imageOpts.SnapChannels[snap]; !found {
-						imageOpts.Snaps = append(imageOpts.Snaps, snap)
-						imageOpts.SnapChannels[snap] = channel
-					}
-				}
-			}
+	// add any extra snaps from the image definition to the list
+	for _, extraSnap := range classicStateMachine.ImageDef.Customization.ExtraSnaps {
+		if !helper.SliceHasElement(classicStateMachine.Snaps, extraSnap.SnapName) {
+			imageOpts.Snaps = append(imageOpts.Snaps, extraSnap.SnapName)
```

Comment:
this can only be done if the model has grade unset or the grade is dangerous. I know that currently only the former is true for classic model, but that will not be true forever